### PR TITLE
Core/adding getradius to base geometry

### PR DESCRIPTION
**Pull request by GuillermoCasas**
I am creating this pull request to start discussing this on more concrete grounds (see #4047). The biggest weakness of the implementation is that the sphere keeps a (raw) pointer to the radius, which must be set from outside. This means that it could potentially fall out of scope, rendering the pointer undefined. This set method is only allowed to be done once in order to avoid erroneous use. In DEM application it would be made to point to the nodal value RADIUS, so that it would automatically change with it. I have added a bunch of tests too that check Get/Set methods for the radius, number of faces and edges and the HasIntersection() function with a box.

Talking to @pooyan-dadvand, he suggested to change the constructor and pass two points instead of one to the sphere and just encode in the second point the info about the radius. However, this is not satisfactory, since it is obscure and also it does not allow one to link the radius to the nodal value (it requires the use of a DELTA_RADIUS in analogy to the DISPLACEMENT that must be added to the initial position to get the current position [X,Y,Z]). But also this complicates reading the mdpa and turns it into a strange-looking list with many repeated values (for constant radii). In summary, the fact is I think that the notion of geometry in Kratos is a bit too tight for a sphere and I could only fit it this way. 

Anyway, please feel free to criticise this solution; any ideas will be most welcome.

### CODE REVIEW

**Comment by loumalouomega** [comment]
Look at the conflict.

Speaking about the implementation, I think i is fine as far as you check if the pointer is NULL.

**Review by RiccardoRossi — COMMENTED** [review]
@KratosMultiphysics/technical-committee

File: kratos/geometries/sphere_3d_1.h [hunk 1]
```diff
@@ -583,25 +631,41 @@ class Sphere3D1 : public Geometry<TPointType>
...
+        return *mpRadius;
+    }
+
+    /** Counterpart of GetRadius(). It provides an interface
+     * to set the radius when it cannot be inferred from other
+     * information in the object (e.g., in a sphere)
+     */
+
+    void SetRadius(const double& rRadius) override {
+        KRATOS_ERROR_IF(mpRadius != NULL) << "Trying to set the sphere's radius for a second time to a value other than NULL. This operation must be done only once." << std::endl;
```
Comment (philbucher) [inline kratos/geometries/sphere_3d_1.h]:
```suggestion
        KRATOS_ERROR_IF_NOT(mpRadius) << "Trying to set the sphere's radius for a second time to a value other than NULL. This operation must be done only once." << std::endl;
```

File: kratos/geometries/sphere_3d_1.h [hunk 2]
```diff
@@ -583,25 +631,41 @@ class Sphere3D1 : public Geometry<TPointType>
...
 
     ShapeFunctionsGradientsType& ShapeFunctionsIntegrationPointsGradients( ShapeFunctionsGradientsType& rResult, IntegrationMethod ThisMethod ) const override
     {
-        std::cout<<"This method (ShapeFunctionsLocalGradients) has no meaning for this type of geometry (Sphere)."<<std::endl;
+        KRATOS_WARNING("Sphere3D1") << "This method (ShapeFunctionsLocalGradients) has no meaning for this type of geometry (Sphere)."<<std::endl;
         return( rResult );
     }
 
+    double GetRadius() const override{
+        KRATOS_ERROR_IF(mpRadius == NULL) << "The Sphere3D1's radius address has not been properly set before calling the method 'GetRadius'. Use 'SetRadius' for that." << std::endl;
```
Comment (philbucher) [inline kratos/geometries/sphere_3d_1.h]:
```suggestion
        KRATOS_ERROR_IF_NOT(mpRadius) << "The Sphere3D1's radius address has not been properly set before calling the method 'GetRadius'. Use 'SetRadius' for that." << std::endl;
```

**Review by philbucher — COMMENTED** [review]
you could also add tests to check if the exeption is thrown if the radius is notinitialized

File: kratos/geometries/sphere_3d_1.h (hunk 2 above)
Comment (GuillermoCasas) [reply kratos/geometries/sphere_3d_1.h]: done. (actually, `KRATOS_ERROR_IF(mpRadius)`).

File: kratos/geometries/sphere_3d_1.h (hunk 1 above)
Comment (GuillermoCasas) [reply kratos/geometries/sphere_3d_1.h]: done.

**Comment by GuillermoCasas** [comment]
> Look at the conflict.
> 

Done.

File: kratos/geometries/geometry.h [hunk 3]
```diff
@@ -2241,6 +2241,27 @@ class Geometry : public PointerVector<TPointType>
...
+    /// Generic geometric methods
+
+    /** It returns the radius of the geometry.
+     * For any geometry G the radius can be defined as
+     * 1/2 de diameter, where
+     * diameter = sup { d(x, y) | x, y ∈ G }
+     * This notion trivially generalizes the usual sphere or circle radii.
+     */
+    virtual double GetRadius() const {
+      KRATOS_ERROR << "Calling base class 'GetRadius' method instead of derived class one. Please check the definition of derived class. " << *this << std::endl;
```
Comment (RiccardoRossi) [inline kratos/geometries/geometry.h:2253]: i would call the CircumRadius here by default

**Comment by philbucher** [comment]
FYI the @KratosMultiphysics/technical-committee discussed this briefly while discussing the integration of the CAD-geometries and maybe there will be a solution that will also work here

**Review by roigcarlo — CHANGES_REQUESTED** [review]
@KratosMultiphysics/technical-committee agrees about having 
a function GetRadius, but 
disagrees about adding a function SetRadius since the usage would be very specific for DEM alone.

if there is no other way around one may eventually use a special geometry for DEM and implement SetRadius for the specific geometry and do an upcasting to call it. What we are trying to express is that within DEM one may treat the pointer to a generic geometry as a pointer to SpecialDEMGeometry or whatever that is called, which allows the specific behaviour.

in any case a "standard" sphere, should not rely on this special behaviour, ad should access the database instead 

(on the line of...)

~~~C++
double GetRadius() const
{
    return (*this)[0].FastGetSolutionStepValue(RADIUS); 
}
~~~

**Comment by loumalouomega** [comment]
> 
> (on the line of...)
> 
> ```c++
> double GetRadius() const
> {
>     return (*this)[0].FastGetSolutionStepValue(RADIUS); 
> }
> ```

This will only work for geometries of nodes, and that's not necessarily true

**Comment by roigcarlo** [comment]
@loumalouomega yes, but if you fall back to the base class it will throw a meaningful error.

also note that if one really needs a pointer than 

~~~cpp
double* pR = &GetGeometry().GetRadius();
~~~

would give back a pointer which is valid within a time step.

**Comment by pooyan-dadvand** [comment]
Giving it another try at @KratosMultiphysics/technical-committee we see several cons to this design:
1. Having a pointer to the radius at least is a very strange design from geometry point of view. 
2. It also gives trouble in memory management. At this moment no one is deleting it.
3. If someone calls a `SetRadius(local_variable_radius)` it will give a segmentation fault
4. The external dependency of the radius (pointing to) is not well justified.

Meanwhile, we agree with having the `GetRadius` and `SetRadius` as virtual methods of the base Geometry class. We suggest having a `double mRadius` in the sphere geometry instead of the pointer. 

The issue of constructor with two points is still open to be discussed.

**Comment by pooyan-dadvand** [comment]
For the constructor we consider following options (also open to other proposals):
1. Having the constructor with only one Node which initialize the Radius to zero or to a non physical value like -1
2. Having an specific constructor with Node and double radius to be used in the future in the new Json format. #6163

**Comment by GuillermoCasas** [comment]
> For the constructor we consider following options (also open to other proposals):
> 
> 1. Having the constructor with only one Node which initialize the Radius to zero or to a non physical value like -1
> 2. Having an specific constructor with Node and double radius to be used in the future in the new Json format. #6163

I think I prefer the second one. This is not urgent, it is only a matter of code coherence and elegance. Therefore, I think it is best to rely on the future json-based definition of the geometry, since the first option is not clean in my opinion. We will discuss this with the other memebers of the @KratosMultiphysics/implementation-committee tomorrow.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/implementation-committee would you please put here your conclusion for the record?

**Comment by pablobecker** [comment]
On behalf of the @KratosMultiphysics/implementation-committee 
We still think it is interesting for the sphere geometry to have a radious defined in it and accessible (not neccesarilty through the node)
We believe this should be possible in the new style of geometry. As soon as there is renewed interested in having this feature, these ideas should be taken into account.

**Pull request closed without merging**